skyportal/handlers/api/source.py

```diff
@@ -946,28 +920,75 @@ class Validator(Schema):
+                    )
+                taxonomy_names, classifications = list(
+                    zip(
+                        *list(
+                            map(
+                                lambda c: (
+                                    c.split(":")[0].strip(),
+                                    c.split(":")[1].strip(),
+                                ),
+                                classifications,
+                            )
+                        )
+                    )
+                )
+                classification_accessible_query = (
+                    Classification.query_records_accessible_by(
+                        self.current_user
+                    ).subquery()
+                )
+
+                classification_query = (
+                    DBSession()
+                    .query(
+                        distinct(Classification.obj_id).label("obj_id"),
+                        Classification.classification,
+                    )
+                    .join(Taxonomy)
+                    .filter(Classification.classification.in_(classifications))
+                    .filter(Taxonomy.name.in_(taxonomy_names))
+                )
+                classification_subquery = classification_query.subquery()
+
+                # We join in the classifications being filtered for first before
+                # the filter for accessible classifications to speed up the query
+                # (this way seems to help the query planner come to more optimal join
+                # strategies)
+                obj_query = obj_query.join(
+                    classification_subquery,
+                    Obj.id == classification_subquery.c.obj_id,
+                )
+                obj_query = obj_query.join(
+                    classification_accessible_query,
+                    Obj.id == classification_accessible_query.c.obj_id,
+                )
+
             else:
-                return self.error(
-                    "Invalid classifications value -- must provide at least one string value"
+                # Not filtering on classifications, but ordering on them
+                classification_query = Classification.query_records_accessible_by(
+                    self.current_user
                 )
-            # Parse into tuples of taxonomy: classification
-            classifications = list(
-                map(
-                    lambda c: (c.split(":")[0].strip(), c.split(":")[1].strip()),
-                    classifications,
+                classification_subquery = classification_query.subquery()
+
+                # We need an outer join here when just sorting by classifications
+                # to support sources with no classifications being sorted to the end
+                obj_query = obj_query.join(
+                    classification_subquery,
+                    Obj.id == classification_subquery.c.obj_id,
+                    isouter=True,
```

Comment:
Wouldn't this be slow compared to what we are currently doing?

Reply:
You may think so, but not in practice. Under the hood, this generates pretty similar queries from the perspective of the Postgres query planner, with similar amounts of nested looping and such. It really only changes the ordering of those loops, like which records/conditions are the innermost loop etc, or something along those lines. I arrived at this through experimentation and a somewhat vague sense of exactly what the query planner is doing to be honest, so I couldn't explain to you exactly why it's faster but it does indeed seem to be the case.

Reply:
At worst I don't think it performs any worse than the previous version (even if it looks a little uglier on the SQLAlchemy side). Definitely, getting rid of the outer join here when not needed is a strictly good thing (and was the idea Ari had originally).